### ISSUE TITLE
Zend incompatibility: yield syntax for sending values to generator

### ISSUE THREAD

**Issue by stuartcarnie** [facebook/hhvm#1627]
The following code fails to compile with hhvm
```php
<?php

function get_gen() {
    echo "Starting generator\n";
    $a = yield;
    echo $a . "\n";
}

$gen = get_gen();
$gen->send(1);
```

Returns (hhvm):
```
HipHop Fatal error: syntax error, unexpected ';' in /home/contatta/tmp/generator_test.php on line 5
```

By adding a simple expression (which is not compatible with Zend), it then complains ->next() must be called, which is also incompatible with Zend:

```php
<?php

function get_gen() {
    echo "Starting generator\n";
    $a = yield 0;
    echo $a . "\n";
}

$gen = get_gen();
$gen->send(1);
```

Returns (hhvm):
```
HipHop Fatal error: Uncaught exception 'Exception' with message 'Need to call next() first' in :
Stack trace:
#0 /home/contatta/tmp/generator_test.php(10): Continuation->send()
#1 {main}
```

The following is compatible with Zend (but not hhvm), and does not require ->next() be called:

```php
<?php

function get_gen() {
    echo "Starting generator\n";
    $a = (yield 0);
    echo $a . "\n";
}

$gen = get_gen();
$gen->send(1);
```

Returns (hhvm):
```
HipHop Fatal error: syntax error, unexpected T_YIELD in /home/contatta/tmp/generator_test.php on line 5
```

Returns (Zend):
```
Starting generator
1
End
```

**Comment by SiebelsTim** [comment]
@paulbiss What happened to this? :)

**Comment by paulbiss** [comment]
Ran into a couple of issues with the way that we guard on stack types and couldn't decide on how it should behave for await. I'll try to get back to this.

**Comment by mtdowling** [comment]
I'm also running into this issue in the AWS SDK for PHP.

`Fatal error: syntax error, unexpected T_YIELD in /home/travis/build/aws/aws-sdk-php/src/ResultPaginator.php on line 68`

**Comment by paulbiss** [comment]
We don't really use the priorities to track issues anymore beyond wishlist generally meaning that no one has plans to work on it. The TL;DR here is I plan to return to this once I reach the end of a long list of other priorities I have unless someone else wants to pick it up.

**Comment by dieend** [comment]
I also run into this problem when using AWS SDK for PHP, when trying to access instance profile credentials..

**Comment by EwanValentine** [comment]
@dieend  Same here!

**Comment by steveh** [comment]
It's a documented issue on AWS http://docs.aws.amazon.com/aws-sdk-php/v3/guide/faq.html#does-the-sdk-work-on-hhvm

Does SDK v2 work?

**Comment by EwanValentine** [comment]
v2 works with HHVM, but v2 doesn't work with Laravel 5.1 unfortunately. I'll have to try the nightly HHVM build or revert back to php-fpm for now

**Comment by steveh** [comment]
@EwanValentine thanks for the info - what's the incompatibility with Laravel and v2?

**Comment by EwanValentine** [comment]
@steveh - Laravel uses ```league/flysystem-aws-s3-v3"``` as its main file storage adapter, which requires v3. Which then throws the compatibility issues with yield

**Comment by jeremeamia** [comment]
This is causing problems for a few different libraries:

* [Guzzle 6](https://github.com/guzzle/guzzle) (and [Guzzle Promises](https://github.com/guzzle/promises))
* [Icicle](https://github.com/icicleio/icicle)
* [AWS SDK for PHP (Version 3)](https://github.com/aws/aws-sdk-php)
* [Recoil](https://github.com/recoilphp/recoil)

 Basically, it makes anything that attempts to do coroutines impossible.

**Comment by peppy** [comment]
This is a pretty serious issue for multiple Laravel packages, as mentioned above. Not sure it should still be low priority.

**Comment by paulbiss** [comment]
From above:
> We don't really use the priorities to track issues anymore beyond wishlist generally meaning that no one has plans to work on it. The TL;DR here is I plan to return to this once I reach the end of a long list of other priorities I have unless someone else wants to pick it up.

I honestly don't know why we haven't gotten rid of that label, I'm getting pretty tired of explain that it's basically meaningless...

**Comment by evert** [comment]
@paulbiss regardless of it being meaningless or not, it would appear that the low-pri label has been pretty accurate given the speed in which this issue is addressed.

18 months and counting.

**Comment by paulbiss** [comment]
Yeah, and I'm really sorry about that. There are only so many of us, and unfortunately no one has the bandwidth to do this right now. This is most certainly on my radar but it's going to be a non-trivial change.

In general the best way to get something prioritized is to send us a patch for it. We absolutely plan to look at this, but I can't drop everything I'm working on right now to do it. While I can understand your frustration this is by no means the oldest issue and I wouldn't characterize it as low-priority. The open source team simply isn't large enough to offer any kind of SLA on issues. Turnaround on reviewing pull-requests, is usually days to weeks, issues (particularly complicated issues) are generally open much longer.

**Comment by evert** [comment]
Totally get that. There's only so much you can do, and I am grateful for what you guys have done. I think the frustration with things like this is not the fact that it's not done yet, but rather the lack of transparency.

Labeling things by priority accurately would be a great help =)

**Comment by jeskew** [comment]
@paulbiss Where would you advise someone writing a patch for this to start looking?

**Comment by paulbiss** [comment]
@jeskew I had a patch some months ago that was still had some bugs to iron out. At this point there's little chance you'd be able to rebase it but I can try to dig it up if you want to use it as a guide.

@evert That's a fair point. We usually try to indicate when we're working on something (comments on the issue, assignments etc), and when we don't ever plan to work on something (wishlist, won't fix), but assigning some priorities could be useful. Part of the problem is that not all of the work we do comes out of the issue queue.

**Comment by jeskew** [comment]
@paulbiss That would be great. Could you email it to jonathan`[at][my username][dot]`net?

**Comment by r3wt** [comment]
aww bum, so much for async uploads to s3

**Comment by paulbiss** [comment]
@jeskew here's the patch as it was the last time I had a chance to work on it: https://reviews.facebook.net/D25821
@r3wt I don't follow? s3?

**Comment by evert** [comment]
@r3wt is just talking about a very specific use-case that is presumably implemented by some code that takes advantage of this feature. Mostly a completely unhelpful comment.

**Comment by fredemmott** [comment]
Amazon's S3 PHP SDK depends on yield behavior that HHVM does not support.

**Comment by Orvid** [comment]
As a note, PHP7 does bring support for the non-parenthesized version, https://3v4l.org/pRDr9

**Comment by rowillia** [comment]
@fredemmott Which parts of yield behavior does HHVM not support?  Once https://github.com/facebook/hhvm/commit/48f0982c35a4ff7ef29088ae62edb62a9974e9d7 and https://github.com/facebook/hhvm/pull/4915 hit we should have all of the behavior we need.  If parsing works then HHVM should work out of the box, right?

**Comment by fredemmott** [comment]
No idea, was just adding more information re @r3wt 's comment.

**Comment by rowillia** [comment]
@fredemmott Ah OK.  We're close though!  I think it'd be pretty amazing for HHVM to support the aws-sdk, it would really help a bunch of startups that want to adopt HHVM :).

**Comment by Jameron** [comment]
I'm really interested in seeing this issue resolved. I moved my entire code base from Laravel 4.2 on Apache (9 Requests Per Second) to Lumen NGinx/FPM (18 RPS), then from FPM to HHVM (100RPS), needed an authentication system back so I moved the code base over to Laravel 5.1(71 RPS), realized AWS-SDK was not working, and went back to NGinx/FPM at around 18 RPS. Would be great to get it back to 71 RPS!

**Comment by fredemmott** [comment]
@scottrice, does your work touch this?

**Comment by jwatzman** [comment]
Some (but I think not all) of this will be solved by @nikic and @paulbiss's generator priming diff, which has been accepted internally and will land real soon now (working through some deployment issues on the FB side, since we rely on the incorrect behavior).

**Comment by shadowhand** [comment]
A lot of new code relies on Guzzle and the AWS-SDK these days. Looking forward to seeing this fixed!

**Comment by jwatzman** [comment]
@shadowhand can you try again with a very recent nightly build? A couple of fixes have landed in the last couple of weeks (more recently than the last stable release), and I'm hopeful this is fixed now, or at least in a better state?

**Comment by bt** [comment]
Not sure if there's been any progress on this issue? This issue is currently a blocker for my application to use HHVM due to the AWS SDK relying on it. :cry:

**Comment by fredemmott** [comment]
All the examples given in the original are fixed in 3.11, probably as part of the PHP7 work:

https://3v4l.org/rFJWQ
https://3v4l.org/pRDr9
https://3v4l.org/0qWQW

If yield-related incompatibilities still exist, please open a separate issue with example code (small, isolated examples like this issue has are ideal).
